### ISSUE TITLE
Queries are waiting and timing out fetching connections when all connections are idle according to metrics preview feature

### ISSUE THREAD

**Issue by danafallon** [prisma/prisma#16314]
### Bug description

We're using prisma client 4.4.0 in production and we just started logging [metrics](https://www.prisma.io/docs/concepts/components/prisma-client/metrics) in order to help diagnose the `Timed out fetching a new connection from the connection pool error` that we see sometimes. I’m looking at the metrics from the time when one of those errors was thrown, and it looks like we consistently had waiting queries, even though all of the connections in the connection pool were idle and no queries were currently running:
```
prisma_client_queries_active 0
prisma_client_queries_wait 12
prisma_pool_connections_busy 0
prisma_pool_connections_idle 20
```
Our connection limit is 20 and we have the default connection timeout of 10s. I’m wondering why there are queries waiting when nothing is currently executing and all of the connections seem to be available.

Two other odd things from the metrics at that same point in time:
1. The metrics contain `prisma_pool_connections_open` twice, and one of them has a negative value (-373, which is interesting because it also says that 373 connections have been closed). The other is very high (393), much higher than the connection pool size limit, and equal to the `prisma_pool_connections_opened_total`. This makes me suspect that the number of currently open connections is not being reported accurately.
2. The histogram of query wait times seems to indicate that no query waited more than 1ms, because all the values are the same. I'm not sure how that's possible when the metrics at this timepoint and many timepoints leading up to it showed that there were waiting queries.

Here's the full output of the metrics at this particular timepoint (in prometheus format):
```ini
# HELP prisma_client_queries_total Total number of Prisma Client queries executed
# TYPE prisma_client_queries_total counter
prisma_client_queries_total 28361
# HELP prisma_datasource_queries_total Total number of Datasource Queries executed
# TYPE prisma_datasource_queries_total counter
prisma_datasource_queries_total 31750
# HELP prisma_pool_connections_closed_total Total number of Pool Connections closed
# TYPE prisma_pool_connections_closed_total counter
prisma_pool_connections_closed_total 373
# HELP prisma_pool_connections_open Number of currently open Pool Connections
# TYPE prisma_pool_connections_open counter
prisma_pool_connections_open 396
# HELP prisma_client_queries_active Number of currently active Prisma Client queries
# TYPE prisma_client_queries_active gauge
prisma_client_queries_active 0
# HELP prisma_client_queries_wait Number of Prisma Client queries currently waiting for a connection
# TYPE prisma_client_queries_wait gauge
prisma_client_queries_wait 12
# HELP prisma_pool_connections_busy Number of currently busy Pool Connections (executing a datasource query)
# TYPE prisma_pool_connections_busy gauge
prisma_pool_connections_busy 0
# HELP prisma_pool_connections_idle Number of currently unused Pool Connections (waiting for the next datasource query to run)
# TYPE prisma_pool_connections_idle gauge
prisma_pool_connections_idle 20
# HELP prisma_pool_connections_open Number of currently open Pool Connections
# TYPE prisma_pool_connections_open gauge
prisma_pool_connections_open -373
# HELP prisma_pool_connections_opened_total Total number of Pool Connections opened
# TYPE prisma_pool_connections_opened_total gauge
prisma_pool_connections_opened_total 396
# HELP prisma_client_queries_duration_histogram_ms Histogram of the duration of all executed Prisma Client queries in ms
# TYPE prisma_client_queries_duration_histogram_ms histogram
prisma_client_queries_duration_histogram_ms_bucket{le="0"} 0
prisma_client_queries_duration_histogram_ms_bucket{le="1"} 212
prisma_client_queries_duration_histogram_ms_bucket{le="5"} 22915
prisma_client_queries_duration_histogram_ms_bucket{le="10"} 26012
prisma_client_queries_duration_histogram_ms_bucket{le="50"} 27783
prisma_client_queries_duration_histogram_ms_bucket{le="100"} 28179
prisma_client_queries_duration_histogram_ms_bucket{le="500"} 28319
prisma_client_queries_duration_histogram_ms_bucket{le="1000"} 28328
prisma_client_queries_duration_histogram_ms_bucket{le="5000"} 28336
prisma_client_queries_duration_histogram_ms_bucket{le="50000"} 28348
prisma_client_queries_duration_histogram_ms_bucket{le="+Inf"} 28361
prisma_client_queries_duration_histogram_ms_sum 1154277.0576979998
prisma_client_queries_duration_histogram_ms_count 28361
# HELP prisma_client_queries_wait_histogram_ms Histogram of the wait time of all Prisma Client Queries in ms
# TYPE prisma_client_queries_wait_histogram_ms histogram
prisma_client_queries_wait_histogram_ms_bucket{le="0"} 0
prisma_client_queries_wait_histogram_ms_bucket{le="1"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="5"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="10"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="50"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="100"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="500"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="1000"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="5000"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="50000"} 28374
prisma_client_queries_wait_histogram_ms_bucket{le="+Inf"} 28374
prisma_client_queries_wait_histogram_ms_sum 41.092736
prisma_client_queries_wait_histogram_ms_count 28374
# HELP prisma_datasource_queries_duration_histogram_ms Histogram of the duration of all executed Datasource Queries in ms
# TYPE prisma_datasource_queries_duration_histogram_ms histogram
prisma_datasource_queries_duration_histogram_ms_bucket{le="0"} 0
prisma_datasource_queries_duration_histogram_ms_bucket{le="1"} 1170
prisma_datasource_queries_duration_histogram_ms_bucket{le="5"} 27289
prisma_datasource_queries_duration_histogram_ms_bucket{le="10"} 29659
prisma_datasource_queries_duration_histogram_ms_bucket{le="50"} 31195
prisma_datasource_queries_duration_histogram_ms_bucket{le="100"} 31570
prisma_datasource_queries_duration_histogram_ms_bucket{le="500"} 31706
prisma_datasource_queries_duration_histogram_ms_bucket{le="1000"} 31716
prisma_datasource_queries_duration_histogram_ms_bucket{le="5000"} 31724
prisma_datasource_queries_duration_histogram_ms_bucket{le="50000"} 31737
prisma_datasource_queries_duration_histogram_ms_bucket{le="+Inf"} 31750
prisma_datasource_queries_duration_histogram_ms_sum 1157139.4216220004
prisma_datasource_queries_duration_histogram_ms_count 31750
```

### How to reproduce

This situation only seems to happen in our Cloud Run (GCP) deployment. I haven't seen it when running prisma locally.

### Expected behavior

I would expect the metrics to say that there are no idle connections and 20 busy connections right before the `Timed out fetching a new connection from the connection pool error` happens. Since that isn't the case, it's not clear to me why it times out trying to fetch a connection.

I would also expect the metrics to say that all connections are busy when there are any queries waiting. If the connections are idle, I'm not sure what the queries are waiting on. I’m curious about the order of operations with these metrics - when does a connection get marked as busy, and when does a query get marked as waiting?

### Prisma information

```prisma
generator client {
  provider        = "prisma-client-js"
  binaryTargets   = ["native", "linux-musl"]
  previewFeatures = ["metrics"]
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}
```

### Environment & setup

- OS: Alpine Linux in Google Cloud Run
- Database: MySQL 5.7 - this is a large database shared with other services (only one of them using prisma client)
- Node.js version: v14.17.3

### Prisma Version

4.4.0 (Prisma Client only)

**Comment by wickdninja** [comment]
we have been experiencing a very similar issue ( we are on postgres ) 

I am about to go into the process of logging out the metrics as you have done.

**Comment by Raynos** [comment]
Were either of you able to find the answer to this problem ? 

I am also getting connection pool timeouts and dont know what to do

**Comment by janpio** [comment]
Can you open a new issue? As you are running in a serverless/function environment, this might be very different @HaydenJH. Thanks.

**Comment by rymcol** [comment]
Also experiencing this

**Comment by aqrln** [comment]
Hey folks, the upcoming Prisma 5.22 release on Tuesday, Nov 5 will include a fix that may solve the metrics side of this issue. In particular, the negative and constantly growing metrics should be now fixed.

As for the connection timeout issue, this upcoming release also contains a fix for one of the common causes (there isn't a single cause that affected all users however). Your case is likely to be fixed if:
1. You are using PostgreSQL.
2. You are using TLS encryption to connect to the database (this is the case by default whenever the database supports it, e.g., on AWS RDS, Supabase, Neon).
3. Your application is running in a resource-constrained environment, e.g., small EC2 instances or Lambda functions.

Please try it out and let us know if it solves one or both problems for you!

**Comment by ludralph** [comment]
Closing this in favour of [23525](https://github.com/prisma/prisma/issues/23525)
